### DISCUSSION TITLE
Alternatives to protocols

### DISCUSSION THREAD

**Discussion by mynomoto** [clj-easy/tools.misc#3] [Ideas]
I started to think about this problem with the idea of using protocols but I'm not sure anymore.

One alternative is to create a library based on Polylith components (@jeroenvandijk suggested that on slack). I like this alternative because we can have all the implementations over the same repository and make sure that they behave functionally the same. If you want you can always add a specific library and use it but you could use only one version of the polylith based library if we are sneaky using versions. (a version could be something like "1.0.0-cheshire" for a json client based on cheshire, "1.0.0-jsonista" and so on for other libraries)

So each implementation would be using the same namespace but each published library would have only one of the implementations because of how we can use polylith. Changing the implementation would be as easy as changing the version.

If one dreams big we could think about catching exceptions and throwing a common exception so user code would not have to change at all when changing implementations.

WDYT?

**Comment by borkdude** [comment]
I don't have a lot of experience with polylith, but I don't want to include any other dependencies to solve this problem: it should also work out of the box in babashka for example. Keeping it bare-bones is my preference.

**Reply by mynomoto** [reply]
It would work out of the box, the artifact is a regular clojure library, poly is a convenient way to have multiple implementations of the same namespace.

**Reply by mynomoto** [reply]
Let me try a more detailed explanation. Poly would be used in the library, this is true. But users could use the library without knowing about poly at all, it is not a dependency.

**Reply by borkdude** [reply]
What's the benefit of using poly over just protocols or multimethods? This library should be very minimal and fast to load (e.g. in bb). Is that achievable when using poly as well?

I must admit that I lack knowledge of poly so an even more detailed explanation could help convince me.

**Reply by borkdude** [reply]
@mynomoto I must admit that I lack knowledge of poly so an even more detailed explanation could help convince me.

**Reply by mynomoto** [reply]
Sorry about that not being very detailed at all.
Poly is a convenience tool, let me try explaining what I mean without poly.

We could have a namespace e.g. `tools-misc.json` with the functions `write-str` and `read-str`. Different projects could implement the same namespace using different libraries like clojure.data.json, cheshire or jsonista. Each jar would have this namespace with one implementation. To avoid the possibility of conflicts all projects could use the same coordinate e.g. `io.github.tools-misc/tools-misc` and having versions like `1.0.0-jsonista` so it is clear what library is being used.

Poly is a convenient way to have all those projects in the same repo. I will not be a dependency for the resulting jar. Let me know if this helps or if something is missing on the explanation.

**Reply by mynomoto** [reply]
Sean Cornfield wrote about having two implementations of an http client on his blog https://corfield.org/blog/2021/10/13/deps-edn-monorepo-7/ on the section Polylith & Swappable Implementations. The idea is publish each implementation as a different library version.

**Reply by borkdude** [reply]
@mynomoto Gotcha. I think that's interesting, but personally, I think this library should not take on the burden of managing implementations: it should mainly take on the burden of providing the mechanism to program against a non-implementation and provide a means to register an implementation. Everything beyond that, pushing implementations to clojars, etc, can be done by third parties.

**Reply by borkdude** [reply]
And yes, we could provide one or two implementations for ease of use, but even then, I think we shouldn't couple it to the specific versions of those libraries (but tests of this library will of course use specific ones).

**Reply by borkdude** [reply]
@mynomoto If I understand the polylith idea, then it would be similar to the below idea, but the difference would be that someone would provide an override for `(ns clj-easy.tools.misc.json)` in their library? That can work, but it could become pretty hard to detect when you would have multiple implementations competing with each other, which I think should lead to a warning. Maybe an implementation should be explicitly selected by requiring a namespace like `cheshire.tools.misc` which then registers itself.

**Reply by mynomoto** [reply]
I think the main point we disagree is that I think this should be centralized and you think it should be decentralized.

The ideia of one namespace having multiple implementations only makes sense imo if this is centralized. Otherwise protocols, multimethods or a way to register seem better options.

I think that for the implementation to be swappable it should have the same behavior including throwing the same exceptions or else we would be setting users up to be surprised. I'm not sure if this is possible/practical but it would be great to have. I will try to write a version of the centralized idea soon to find if it is a crazy idea at least.

**Reply by borkdude** [reply]
Wether or not it's centralized or not: could be, I'm not disagreeing necessarily, but ultimately, like say with clj-kondo configs/hooks, the maintenance of implementations belong to the library authors, not the project that defines the interface. So there should also be a way to provide implementations "out of band".

**Reply by borkdude** [reply]
Of course the interfaces should be good enough to accommodate several libraries, that's where I see the most benefit of test-driving the implementations, in addition to providing them for others.

**Reply by borkdude** [reply]
Providing an out of band implementation for clojure.tools.logging is done by timbre as follows:

https://github.com/ptaoussanis/timbre/blob/master/src/taoensso/timbre/tools/logging.clj

What I like about that approach: 

Using timbre as a tools logging provider is explicit. There is absolutely no confusion which provider "wins", compared to the "who is first on the classpath" approach. 

```
(taoensso.timbre.tools.logging/use-timbre)
```

What I currently don't like about the specific mechanism in tools.logging: 

- the reliance on alter-var-root: this could give problems with direct linking and also isn't cross-platform (CLJS). Using an explicit registration function (so the internals of the "framework" can decide how functions get registered) seems an improvement.

- usage of protocols leads to either `reify` or `defprotocol`. We could just use maps and functions. An explicit registration function can do some sanity checks on the map with functions.

Both of these problems are addressed in the below approach:

https://github.com/clj-easy/tools.misc/discussions/3#discussioncomment-2492288

**Reply by borkdude** [reply]
Another point: as long as the registration function is explicit, it doesn't even matter if we use protocols or whatever, this can be decided on a case by case basis I guess: whatever makes sense for the tool at hand.

**Comment by borkdude** [comment]
Another approach could just be to not use protocols or multimethod but a light-weight plugin system.

This library defines functions that delegate to an implementation. 

``` clojure
(ns clj-easy.tools.misc.json)

(def ^:private write-str-impl (volatile! nil))

(defn write-str [s] (@write-str-impl s))

(defn register-json [{:keys [write-str]}]
  (vswap! write-str-impl write-str))
```

When loading an implementation, it can register itself:

``` clojure
(ns clj-foobar.library (:require [clj-easy.tools.misc.json :as tmj] [cheshire.core]))
(tmj/register-json {:write-str cheshire.core/generate-str})
```

The `clj-foobar.library` namespace registers cheshire as the implementation. We could provide this namespace so it can be used by app users and library users for testing only.

We could generalize this to a keywordized system, like `(reg! :clj-easy.tools.misc.json/write-str cheshire/generate-str)` as well.

We could also warn and check if multiple libraries register implementations for the same thing, which imo should not happen in practice (since then one of the libraries on the classpath does not play by the rules that only end-users should provide implementations).

**Comment by borkdude** [comment]
I agree in general that we should avoid using protocols on the API surface. The API surface of each tool can just be a registration function. I've committed that as an example. The tests now look like:

``` clojure
(ns clj-easy.tools.misc.json-test
  (:require [clj-easy.tools.json :as json]
            [clj-easy.tools.json.cheshire :as tools.cheshire]
            [clj-easy.tools.json.data-json :as tools.data.json]
            [clojure.test :refer [deftest is testing]]))

(tools.cheshire/reg-provider!)
(tools.data.json/reg-provider!)

(deftest read-str-test
  (doseq [provider '[cheshire/cheshire org.clojure/data.json]]
    (json/choose-provider! provider)
    (testing "read json"
      (is (= [1 2 3] (json/read-str "[1, 2, 3]")))
      (is (= {:a 1} (json/read-str "{\"a\": 1}")))
      (is (= {:a 1} (json/read-str "{\"a\": 1}" {:key-fn keyword})))
      (is (= {"a" 1} (json/read-str "{\"a\": 1}" {:key-fn str}))))
    (testing "write json"
      (is (= [1 2 3] (json/read-str (json/write-str [1 2 3]))))
      (is (= {:a 1} (json/read-str (json/write-str {:a 1})))))))
```

Note that there is no problem with having multiple providers on the classpath: registration and choosing a provider is explicit. 
When choosing a provider while there was already one chosen previously, a warning is emitted:

``` clojure
$ clojure -M:test

Running tests in #{"test"}

Testing clj-easy.tools.misc.json-test
WARNING: json provider was already set to: cheshire/cheshire and is now set to: org.clojure/data.json

Ran 1 tests containing 2 assertions.
0 failures, 0 errors.
```

Thoughts, @mynomoto ?

**Comment by bsless** [comment]
Coming over from other discussions I've opened, I have an objection to _registering_. I can see it blow up in my face in some creative way in the future, the joys of side effects.
Alternative solution? Forgive me for uttering the phrase, but "factory"?
`(def parser (json/provider :cheshire)` where the provider returns a function.
I'm not a fan of global registries and definitely wouldn't want to introduce a volatile read on the code path for no reason.

**Reply by borkdude** [reply]
Sure, there are other ways. The volatile is an implementation detail that could be changed. Where would `(json/provider :cheshire)` look up its implementation if there wasn't some kind of registry?

**Reply by borkdude** [reply]
And don't tell me multimethods, since they are effectively a global registry too :)

**Reply by bsless** [reply]
That was just an example, not a thought-out api.
A ~good~ better api will probably be:
`(json/provider {:read (fn [,,] ,,) :write ,,})`

**Reply by borkdude** [reply]
I don't understand where is that API used from and how. What does `json/provider` do? I think it's not a very good UX if this means you would have to define the `read` and `write` vars yourself again to wrap those functions. Since how can libraries program against a contract this way?

**Reply by jeroenvandijk** [reply]
The current implementation for json looks good to me. Since parsing json and writing json is stateless it is unlikely that you have to switch the provider more than once (in a normal program). 

For http things are more complicated as it is more likely you will swap the implementation and it needs to be threadsafe (e.g. in test environments of the users)

Some remarks:

- Maybe the read and write function could be wrapped in a function that throws when it is nil to prevent a cryptic nullpointer exception when it is not set
- Should the provider be selected the moment you register it, so it is only one step of requiring the namespace? I wonder what the use case is of choosing a provider other than in test environments?
- Choosing a provider could also have an extra check if the provider exists to catch errors early on

**Reply by borkdude** [reply]
It looks like you're reacting to @bsless's comment here, but assuming you are talking about the current JSON impl: good points. 

I think it's up to the library that offers the implementation if it will register on namespace require, we can certainly do so for our default implementations, so it will be one step less.
You could even "choose-provider" automatically on namespace require. Then the require of the impl namespace does both things, so it will all be automatic. 

The NPEs: sure that's easy to solve. You can even check more things in the "choose-provider" function, e.g. the existence of certain keys.

**Reply by jeroenvandijk** [reply]
Yeah sorry, indeed about the [current implementation](https://github.com/clj-easy/tools.misc/blob/78568fb20b04a8beecdf4a8c5a7c19864c40ae26/src/clj_easy/tools/json.clj#L10). You seemed to have continued after the discussion with @bsless, but it was a bit confusing. And I still need to get used to this discussion interface as well :)

Agree about the default implementations. And more specific checks will be helpful indeed.

**Reply by borkdude** [reply]
@bsless commented about the use of volatile!: 

I think in pretty much all cases I can think of: JSON, HTTP Requests, etc, reading a volatile is neglectible compared to the rest of the work done in that function. An alternative could be using alter-var-root but that doesn't play well with direct linking.

@mynomoto is trying out a polylith-like approach where there is only one `clj-easy.tools.json` namespace which defines the functions and it depends on your dependency on the classpath who brings that namespace. That could also work (and there would be no intermediate volatile, registration, etc) but could become confusing when you have problems with dependencies. The system I have currently for JSON allows having multiple providers on the classpath and choosing them explicitly.
